elastic/eui
# build: update react-router-dom and cleanup related dependencies usage

## Summary

In order to make react-router-dom compatible with React 18 as well as previous versions we need to use the latest react-router-dom v5. Additionally this also fixes deprecated or legacy API usage.

Changes:

- Update `react-router-dom` to 5.3.4
- Remove `react-router` as docs portal dependency and replace imports with `react-router-dom` imports
- Remove `react-router-redux` because it isn't working since react-router v4 and hasn't been used in our code base in 3 years
- Replace duplicated (and pretty much unused) `createHashHistory` with `<HashRouter>`
- Replace `withRouter` with react-router hooks

This PR resolves #6891 and is related to React 18 support work but can be merged directly into `main` as it doesn't cause any errors or incompatibilities.

## QA

1. Make sure local and CI builds are passing
2. Make sure unit and e2e tests are passing
3. Make sure the application works as previously and there are no regressions related to routing

### General checklist
